sdk/samples/iot/pnp/pnp_protocol.c

```diff
@@ -30,125 +30,111 @@ static az_span const iot_hub_twin_desired_version = AZ_SPAN_LITERAL_FROM_STR("$v
-          return AZ_ERROR_UNEXPECTED_CHAR;
-        }
+        IOT_SAMPLE_EXIT_IF_FAILED(az_json_reader_next_token(jr), log);
         continue;
       }
 
+      // Found a property.
       az_json_token property_name = jr->token;
-
-      if (az_result_failed(az_json_reader_next_token(jr)))
-      {
-        IOT_SAMPLE_LOG_ERROR("Failed to get next token.");
-        return AZ_ERROR_UNEXPECTED_CHAR;
-      }
+      IOT_SAMPLE_EXIT_IF_FAILED(az_json_reader_next_token(jr), log);
 
       property_callback(component_name, &property_name, *jr, version, context_ptr);
     }
-
-    if (jr->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    else if (jr->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
     {
-      if (az_result_failed(az_json_reader_skip_children(jr)))
-      {
-        IOT_SAMPLE_LOG_ERROR("Failed to skip children of object.");
-        return AZ_ERROR_UNEXPECTED_CHAR;
-      }
+      IOT_SAMPLE_EXIT_IF_FAILED(az_json_reader_skip_children(jr), log);
     }
     else if (jr->token.kind == AZ_JSON_TOKEN_END_OBJECT)
     {
       break;
     }
   }
-
-  return AZ_OK;
 }
 
 // Move reader to the value of property name.
-static az_result json_child_token_move(az_json_reader* jr, az_span property_name)
+static bool json_child_token_move(az_json_reader* jr, az_span property_name)
 {
+  iot_sample_error_log log;
+  iot_sample_error_log_init(
+      &log, "Failed to process device twin message: az_result return code 0x%08x.", AZ_SPAN_EMPTY);
+
   while (az_result_succeeded(az_json_reader_next_token(jr)))
   {
     if ((jr->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
-        && az_json_token_is_text_equal(&(jr->token), property_name))
+        && az_json_token_is_text_equal(&jr->token, property_name))
     {
-      if (az_result_failed(az_json_reader_next_token(jr)))
-      {
-        IOT_SAMPLE_LOG_ERROR("Failed to get next token.");
-        return AZ_ERROR_UNEXPECTED_CHAR;
-      }
-
-      return AZ_OK;
+      IOT_SAMPLE_EXIT_IF_FAILED(az_json_reader_next_token(jr), log);
+      return true;
     }
     else if (jr->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
     {
-      if (az_result_failed(az_json_reader_skip_children(jr)))
-      {
-        IOT_SAMPLE_LOG_ERROR("Failed to skip child of complex object.");
-        return AZ_ERROR_UNEXPECTED_CHAR;
-      }
+      IOT_SAMPLE_EXIT_IF_FAILED(az_json_reader_skip_children(jr), log);
     }
     else if (jr->token.kind == AZ_JSON_TOKEN_END_OBJECT)
     {
-      return AZ_ERROR_ITEM_NOT_FOUND;
+      return false;
     }
   }
 
-  return AZ_ERROR_ITEM_NOT_FOUND;
+  return false;
 }
 
 // Check if the component name is in the model
-static az_result is_component_in_model(
+static bool is_component_in_model(
     az_span component_name,
     az_span const** components_ptr,
     int32_t components_num,
     int32_t* out_index)
 {
-  int32_t index = 0;
-
   if (az_span_size(component_name) == 0)
   {
-    return AZ_ERROR_UNEXPECTED_CHAR;
+    return false;
```

Comment:
`IOT_SAMPLE_EXIT_IF_FAILED` here?

Reply:
No, because if the component is not in the model, there is separate logic that needs to be executed upon return.